### ISSUE TITLE
Why the custom JSON library?

### ISSUE THREAD

**Issue by jeaye** [ipkn/crow#59]
A custom JSON library feels out of place in crow. Was it specifically for performance? C++11/14 features? Header only?

It makes more sense that crow would focus on being a web framework and look to other libraries for providing JSON, especially considering the number of JSON libraries for C++. Furthermore, it would be comforting to use a JSON library which has extensive unit tests.

To help make my point, here's a list of (some) C++ JSON libraries (one of which is my own): https://github.com/fffaraz/awesome-cpp#json

**Comment by ipkn** [comment]
I include JSON and mustache with crow because of the "Batteries included" philosophy.

To me, giving a default option is important. I can delay decisions like which json library to use and concentrate on the project. 

If someone really need a performance (or their specific objectives), he or she would compare choices and choose accordingly. You can use your json library with crow without any problem. Crow works without embedded json or mustache (just one helper function exists for easy response). That's why I put json and mustache into embedded namespace, `crow::json` and `crow::mustache`.

**Comment by jeaye** [comment]
> I include JSON and mustache with crow because of the "Batteries included" philosophy.

Copying in an existing JSON library would give you the same benefit.

> To me, giving a default option is important. I can delay decisions like which json library to use and concentrate on the project.

Except for the fact that you have to take time away from the project to write almost 1500 lines of code which you now need to maintain. I mentioned before that the existing implementation lacks unit tests; it's also not very clean to use, compared to the more modern alternatives (which, for example, support initializer lists very well).

1. You have put in a lot of work creating this JSON library
2. The library has no tests
3. The library wasn't necessarily written for speed or safety, but instead just for convenience
4. The library has almost nothing to do with crow and, as you said, stands on its own
5. The library doesn't follow the most modern C++ practices, making it harder to use (and less convenient!)

I'm not saying crow shouldn't ship with a JSON library. I'm just saying the JSON library it currently uses should be swapped out with something maintained by someone else which has tests and has a cleaner interface. For this, I would suggest https://github.com/nlohmann/json (which is not my own), since it hits all of the above points.

Since it makes sense (and even embellishes my point) that you wouldn't want to focus on reworking crow's JSON library while there are other more important/relevant issues and tasks, would you be willing to merge in a PR which switched out the existing JSON library with something else?

**Comment by acron0** [comment]
I agree with @jeaye on this. The JSON parser is the only 'home grown' ancilliary part of Crow. I'd support a PR with an alternative.

**Comment by floatdrop** [comment]
:+1: for https://github.com/nlohmann/json

Some guys say that this code - https://github.com/ipkn/crow/blob/master/include/json.h#L98-L131 is broken and will not behave like it should. I don't know much about C++, but this kind of make me worry.

**Comment by keonkim** [comment]
I agree with @jeaye . I'd make a fork of crow that uses https://github.com/nlohmann/json in the future.

**Comment by unikoid** [comment]
JFYI,

I made some benchmarks (for my use cases only) comparing crow::json to nlohmann/json and the latter one seems to be about two times slower. 
If someone is really interested, I can show you benchmark sources. Note that I'm not a C++ programmer, so my benchmarks can be incorrect.

**Comment by KrishnaPG** [comment]
nlohmann/json is more about syntactic sugar (of C++11) than performance. For raw performance, other json libraries from https://github.com/fffaraz/awesome-cpp#json, such as RapidJson are good. Especially, given their SSE2/SSE4.2 support, small footprint, unicode support etc.

No point in building a server in C++, just to get slow on JSON. For that matter, one should make the whole JSON part as pluggable with other competitor formats, such as Flatbuffers and MsgPack.

I just did `include "crow.h"` and started to get the below prompts from compiler (MSVC):
````
>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xutility(2229): warning C4996: 'std::_Copy_impl': Function call with parameters that may be unsafe - this call relies on the caller to check that the passed values are correct. To disable this warning, use -D_SCL_SECURE_NO_WARNINGS. See documentation on how to use Visual C++ 'Checked Iterators'
1>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xutility(2216): note: see declaration of 'std::_Copy_impl'
1>  \3rdparty\crow\include\json.h(548): note: see reference to function template instantiation '_OutIt std::copy<crow::json::rvalue*,crow::json::rvalue*>(_InIt,_InIt,_OutIt)' being compiled
1>          with
1>          [
1>              _OutIt=crow::json::rvalue *,
1>              _InIt=crow::json::rvalue *
1>          ]
````
